File: September 2022/Day 18/C++ Code/42. Trapping Rain Water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int n = height.size();
        vector<int> left(n,0);
        vector<int> right(n,0);
        int mx =0;
        for(int i=0;i<n;i++) {
            mx = max(mx,height[i]);
            left[i]=mx;
        }
        mx=0;
        for(int i=n-1;i>=0;i--) {
            mx = max(mx,height[i]);
            right[i]=mx;
        }
        int ans=0;
        for(int i=0;i<n;i++) {
            ans+=(min(left[i],right[i])-height[i]);
        }
        return ans;
    }
};

File: September 2022/Day 18/JAVA Code/42. Trapping Rain Water.java
class Solution {
    public int trap(int[] height) {
        int n = height.length;
        int[] left = new int[n];
        int[] right = new int[n];
        int max =0;
        for(int i=0;i<n;i++) {
            max = Math.max(max,height[i]);
            left[i]=max;
        }
        max=0;
        for(int i=n-1;i>=0;i--) {
            max = Math.max(max,height[i]);
            right[i]=max;
        }
        int ans=0;
        for(int i=0;i<n;i++) {
            ans+=(Math.min(left[i],right[i])-height[i]);
        }
        return ans;
    }
}
